src/skagent/model_analyzer.py

```diff
@@ -0,0 +1,274 @@
+                    or (var in self.observables),
+                }
+
+        # 4) reward
+        for blk in self._blocks:
+            for var, rd in blk.reward.items():
+                agent = (
+                    rd
+                    if isinstance(rd, str)
+                    else getattr(rd, "agent", "global") or "global"
+                )
+                self.node_meta[var] = {
+                    "kind": "reward",
+                    "agent": agent,
+                    "plate": agent if agent != "global" else None,
+                    "observed": True,
+                }
+
+        # 5) params: only keys in calibration not appearing on any LHS
+        for p in set(self.calibration) - lhs:
+            self.node_meta.setdefault(
+                p,
+                {"kind": "param", "agent": "global", "plate": None, "observed": False},
+            )
+
+    def _rhs_symbols(self, rule):
+        """Extract variable names from RHS rule (callable or string)."""
+        if isinstance(rule, str):
+            return _TOKEN_RE.findall(rule)
+        if callable(rule):
+            try:
+                return list(inspect.signature(rule).parameters.keys())
+            except Exception:
+                src = inspect.getsource(rule)
+                return _TOKEN_RE.findall(src)
+        return []
+
+    def _collect_dependencies(self):
+        """Collect raw dependencies and build interim param_deps."""
+        for blk in self._blocks:
+            # shock parameters → shock var
+            for var, sd in blk.shocks.items():
+                if isinstance(sd, tuple) and len(sd) == 2:
+                    _, params = sd
+                    if isinstance(params, dict):
+                        for v in params.values():
+                            if isinstance(v, str):
+                                self._raw_deps[var].append(v)
+                                self._param_deps[v].add(var)
+
+            # dynamics
+            for var, rule in blk.dynamics.items():
+                deps = (
+                    list(rule.iset)
+                    if isinstance(rule, Control)
+                    else self._rhs_symbols(rule)
+                )
+                for d in deps:
+                    self._raw_deps[var].append(d)
+                    if d in self.calibration:
+                        self._param_deps[d].add(var)
+
+            # reward
+            for var, rd in blk.reward.items():
+                deps = self._rhs_symbols(rd) if callable(rd) else []
+                for d in deps:
+                    self._raw_deps[var].append(d)
+                    if d in self.calibration:
+                        self._param_deps[d].add(var)
+
+        # dedupe
+        for tgt in list(self._raw_deps):
+            self._raw_deps[tgt] = sorted(set(self._raw_deps[tgt]))
+
+    def _identify_time_dependencies(self):
+        """Mark (tgt, src) pairs for lag edges via insertion-order heuristic."""
```

Comment:
Is this heuristic guaranteed to be accurate? Are there any edge cases that it misses?